authenticatorbridge/src/main/java/com/bitwarden/authenticatorbridge/manager/AuthenticatorBridgeManagerImpl.kt

```diff
@@ -0,0 +1,195 @@
+ * @param symmetricKeyStorageProvider Provides access to local storage of the symmetric encryption
+ * key.
+ * @param callbackProvider Provides a way to construct a service callback that can be mocked in
+ * tests.
+ * @param processLifecycleOwner Lifecycle owner that is used to listen for start/stop
+ * lifecycle events.
+ */
+internal class AuthenticatorBridgeManagerImpl(
+    private val connectionType: AuthenticatorBridgeConnectionType,
+    private val symmetricKeyStorageProvider: SymmetricKeyStorageProvider,
+    callbackProvider: AuthenticatorBridgeCallbackProvider = StubAuthenticatorBridgeCallbackProvider(),
+    context: Context,
+    processLifecycleOwner: LifecycleOwner = ProcessLifecycleOwner.get(),
+) : AuthenticatorBridgeManager {
+
+    private val applicationContext = context.applicationContext
+
+    /**
+     * Main AuthenticatorBridgeService access point.
+     */
+    private var bridgeService: IAuthenticatorBridgeService? = null
+
+    /**
+     * Internal state of [accountSyncStateFlow].
+     */
+    private val mutableSharedAccountsStateFlow =
+        MutableStateFlow<AccountSyncState>(AccountSyncState.Loading)
+
+    /**
+     * Callback registered with AuthenticatorBridgeService.
+     */
+    private val authenticatorBridgeCallback = callbackProvider.getCallback(::onAccountsSync)
+
+    /**
+     * Service connection that listens for connected and disconnected service events.
+     */
+    private val bridgeServiceConnection = object : ServiceConnection {
+        override fun onServiceConnected(name: ComponentName, service: IBinder) {
+            onServiceConnected(service)
+        }
+
+        override fun onServiceDisconnected(name: ComponentName) {
+            onServiceDisconnected()
+        }
+    }
+
+    override val accountSyncStateFlow: StateFlow<AccountSyncState> =
+        mutableSharedAccountsStateFlow.asStateFlow()
+
+    init {
+        // Listen for lifecycle events
+        processLifecycleOwner.lifecycle.addObserver(
+            object : DefaultLifecycleObserver {
+                override fun onStart(owner: LifecycleOwner) {
+                    bindService()
+                }
+
+                override fun onStop(owner: LifecycleOwner) {
+                    unbindService()
+                }
+            },
+        )
+    }
+
+    private fun bindService() {
+        val intent = Intent().apply {
+            component = ComponentName(
+                connectionType.toPackageName(),
+                AUTHENTICATOR_BRIDGE_SERVICE_CLASS,
+            )
+        }
+
+        val isBound = try {
+            applicationContext.bindService(
+                intent,
+                bridgeServiceConnection,
+                Context.BIND_AUTO_CREATE,
+            )
+        } catch (e: SecurityException) {
+            unbindService()
+            false
+        }
+
+        if (!isBound) {
+            mutableSharedAccountsStateFlow.value = AccountSyncState.Error
```

Comment:
We'll need to think about what this particular error would mean for the UI....if you don't have the other app installed, there's no reason to expect this to work. It's almost like we need to do a check to see if the application exists before we even try to bind. If not, we should show something like `AccountSyncState.NoApplication` or something.

Reply:
285c4bafac0195b5ed129df0b3435e87cbd844a5